### ISSUE TITLE
Renaming packages makes environment UI unusable

### ISSUE THREAD

**Issue by healther** [spack/spack#13597]
After the renaming in #13411 using a previously installed environment doesn't work anymore, and I get `py-jupyter-notebook not found`. I think that is to be expected, however the fix requires a manual removal of `spack.lock` and a fixup in `spack.yaml`. At least the first one spack should be able to handle on its own and it should help me find the error in my environment specification

```
# anything before 13411
% spack create general
% spack add py-jupyter-notebook
% spack install
# update to after 13411
% spack env activate general
==> Error: Package 'py-jupyter-notebook' not found.
```

**Comment by adamjstewart** [comment]
> a previously installed environment doesn't work anymore

I think we can do better than this. @alalazo and I had a talk a while back about making sure Spack uses the `packages.yaml` of the package when it is installed, not the current `packages.yaml`. This sounds like a similar case.

In my opinion, package renames are inevitable, and we should fix Spack to handle them better as opposed to fearing package renames.

**Comment by alalazo** [comment]
> I think we can do better than this. @alalazo and I had a talk a while back about making sure Spack uses the packages.yaml of the package when it is installed, not the current packages.yaml. This sounds like a similar case

Which reminds me we have already in the code base a nice case of a repository constructed on the fly in a temporary folder - the one that is used under the hood to concretize specs together. I need to work out the details, but the strategy could be simply to construct a similar repository from the files stored under the `.spack` folder in the prefix of an already installed spec.

**Comment by healther** [comment]
> In my opinion, package renames are inevitable, and we should fix Spack to handle them better as opposed to fearing package renames.

Agreed, although we might have to think about an option to allow for aliases.

Personally I'd be happy with `spack` being smart enough to tell when it fails to even startup in an environment and (as part of the exception) offers me an indication on which files I need to edit/delete as at this point I can't use `spack` anymore to investigate or even leave the environment anymore

**Comment by aweits** [comment]
Just ran into this issue without the complication of the environment activation. Renamed package is installed, but just won't load (find works). This alleviates the issue.

```diff --git a/lib/spack/spack/modules/common.py b/lib/spack/spack/modules/common.py
index 434d09c1254d..51454c01cac8 100644
--- a/lib/spack/spack/modules/common.py
+++ b/lib/spack/spack/modules/common.py
@@ -323,7 +323,11 @@ class UpstreamModuleIndex(object):
 
 
 def get_module(module_type, spec, get_full_path):
-    if spec.package.installed_upstream:
+    try:
+        upstream = spec.package.installed_upstream
+    except spack.repo.UnknownPackageError:
+        upstream, record = spack.store.db.query_by_spec_hash(spec.dag_hash())
+    if upstream:
         module = spack.modules.common.upstream_module_index.upstream_module(
             spec, module_type)
         if get_full_path:```
